### PR TITLE
[Feature][server,api,common,dao,remote] delay execution of tasks and improve some designs

### PR DESCRIPTION
**Pull request by vanilla111**
## What is the purpose of the pull request

- Added a function to delay the execution of tasks;
- Add corresponding unit test;
- Added statistics on delayed execution of tasks;
- Change the time point when the start time of various tasks is set;
- Fix the bug that `switch...case` statement in `NettyDecoder.java` lacks `break`;
- Fix the error that `@ContextConfiguration` lacks `CuratorZookeeperClient.class` in multiple test files;
- Modify the newline format of the `DateUtils.java` file
- The `getRemainTime` method is added to `DateUtils`, and the code for calculating the remaining time is changed to use the new tool method.

## Related issues

- [Feature][Server] Allow tasks to be delayed[#3426](https://github.com/apache/incubator-dolphinscheduler/issues/3426#issue-674696165)
- [Improvement][Server] Time setting [#3352](https://github.com/apache/incubator-dolphinscheduler/issues/3352#issue-668271854)

## Brief change log

- `dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/TaskCountDto.java` added statistics of delayed execution tasks and fix a spelling error;
- `dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/enums/ExecutionStatus.java` added delayed execution status;
- `dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/model/TaskNode.java` added a delay attribute `delayTime`;
- `dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/entity/TaskInstance.java` added delay property `delayTime` and first submission time `firstSubmitTime`;
- Fix the bug that the decode method lacks break in `dolphinscheduler-remote/src/main/java/org/apache/dolphinscheduler/remote/codec/NettyDecoder.java`;
- `dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/builder/TaskExecutionContextBuilder.java` added delay property `delayTime` and current task status property `currentExecutionStatus`;
- Add unit test `dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThreadTest.java `;
- The `createTaskInstance` method in `dolphinscheduler-serve/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterExecThread.java` removes the setting of the start time;
- Condition node execution thread and dependent node execution thread independently set the start time;
- Added judgments on the delayed execution status of tasks `DELAY_EXECUTION` in multiple files;
- `dolphinscheduler-serve/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java` added delayed execution code;
- Add `CuratorZookeeperClient.class` to `@ContextConfiguration` in multiple test files;
- Add fields `delay_time` and `first_submit_time` in the database `t_ds_task_instance` table, and modify the corresponding sql file;
- IDE tool automatically optimizes import **(now, I had use the DS own code style configuration file to format the package import)**

## Verify this pull request

- All specified unit tests pass;
- Passed the black box test;

Here are the details of the black box test:

### test environment

- Operating system: macOS
- Number of Master Nodes: 2
- Number of Worker Nodes: 2
- Worker execution task type: Python

### Test details

The following test temporarily modified the judgment of no failover under the same HOST. The node crashes all use the `-9` semaphore.

Note: If a task needs to be executed later, the WORKER will change the status of the task to `DELAY_EXECUTION` when sending ACK for the first time; after the delay ends, before starting execution, it will send a second ACK to change the status of the task to `RUNNING_EXECUTION` and set the start time. There is only one ACK for tasks that do not need to be delayed.

##### Simple single task

Test case | Master status | Worker status | Expected result | Test result | Description
:-: | :-: | :-: | :-: | :-: | :-:
Dependent node | Normal | Normal | The dependency is handled correctly, the start and end time are correct | As expected
Condition node | Normal | Normal | The condition relationship is processed correctly, the start and end time are correct | As expected
Subprocess node | Normal | Normal | Subprocess execution is Normal, start and end time are correct | As expected
Python node | Normal | Normal | execute the task correctly, start and end time is correct | As expected

#### Task delay execution

Test case | Master status | Worker status | Expected result | Test result | Description
:-: | :-: | :-: | :-: | :-: | :-:
Python node, delay execution for one minute, and the task content is sleep for one minute | Normal | Normal | execute the task correctly, query the task status at each time point in accordance with the real situation | As expected | SUMITED\_SUCCESS -> DELAY\_EXECUTION -> RUNNING\_EXECUTION -> SUCCESS status changed correctly

#### WORKER node crash test

Test case | Master status | Worker status | Expected result | Test result | Description
:-: | :-: | :-: | :-: | :-: | :-:
Python node, delay execution for one minute, and the task content is sleep for one minute | Normal | The node receiving the task crashes before the first ACK | The host is empty, it will not trigger the failover | As expected | At this point Restart the Master, the task is resubmitted
Same as above | Normal | The node that receives the task crashes before the second ACK | Triggers failover | As expected | The master submits a new same task
Same as above | Normal | The node that receives the task crashes while executing the task | Triggers failover | As expected | The master submits a new same task

#### MASTER node crash test

Test case | Master status | Worker status | Expected result | Test result | Description
:-: | :-: | :-: | :-: | :-: | :-:
Python node, delay execution for one minute, and the task content is sleep for one minute | The node dispatching the task crashes before the first ACK | Normal | Another node takes over the task until the task is completed | As expected | 
Same as above | The node that dispatches the task crashes before the second ACK | Normal | Another node takes over the task | As expected | 
Same as above | The node that dispatches the task crashes during the execution of the task | Normal | Another node takes over the task | As expected |

In the above test, the various times of the workflow and task instances are set at the correct time.

### CODE REVIEW

File: dolphinscheduler-remote/src/main/java/org/apache/dolphinscheduler/remote/codec/NettyDecoder.java [hunk 1]
```diff
@@ -55,15 +55,18 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) t
...
                 commandHeader.setType(in.readByte());
                 checkpoint(State.OPAQUE);
+                break;
             case OPAQUE:
                 commandHeader.setOpaque(in.readLong());
                 checkpoint(State.BODY_LENGTH);
             case BODY_LENGTH:
                 commandHeader.setBodyLength(in.readInt());
                 checkpoint(State.BODY);
+                break;
```
Comment (CalvinKirs) [inline dolphinscheduler-remote/src/main/java/org/apache/dolphinscheduler/remote/codec/NettyDecoder.java:69]: You cannot directly break here, please refer to #3035

Comment (CalvinKirs) [reply dolphinscheduler-remote/src/main/java/org/apache/dolphinscheduler/remote/codec/NettyDecoder.java:69]: @gabrywu gabrywu gave some suggestions, maybe you can complete it. You can refer to https://github.com/apache/incubator-dolphinscheduler/pull/2969

Comment (vanilla111) [reply dolphinscheduler-remote/src/main/java/org/apache/dolphinscheduler/remote/codec/NettyDecoder.java:69]: > You cannot directly break here, please refer to #3035

I know the reason for removing break. Should I add a break at the last case? If don't do this, there will be many WARN logs that shouldn't appear in the log.

File: dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/entity/TaskInstance.java [hunk 2]
```diff
@@ -84,6 +82,12 @@
      */
     private ExecutionStatus state;
 
+    /**
+     * task first submit time
+     */
+    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss",timezone="GMT+8")
```
Comment (yangyichao-mango) [inline dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/entity/TaskInstance.java:88]:
```suggestion
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss", timezone="GMT+8")
```
Please configure `checkstyle.xml` and `code style xml` in style directory. After configure that, you can use `option+command+L` to automatic format the code.

Checkstyle.xml configuration lesson: https://www.cnblogs.com/wanshi1989/p/5478050.html
code style xml configuration lesson: https://blog.csdn.net/qq_34579060/article/details/80100040

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java [hunk 3]
```diff
@@ -256,4 +272,60 @@ private void downloadResource(String execLocalPath,
...
+    private void delayExecutionIfNeeded() {
+        long remainTime = DateUtils.getRemainTime(taskExecutionContext.getFirstSubmitTime(),
+                taskExecutionContext.getDelayTime() * 60L);
+        logger.info("delay execution time: {} s", remainTime < 0 ? 0 : remainTime);
+        if (remainTime > 0) {
+            try {
+                Thread.sleep(remainTime * Constants.SLEEP_TIME_MILLIS);
+            } catch (Exception e) {
+                logger.error("exception", e);
+                logger.error("delay task execution failure, the task will be executed directly." +
```
Comment (yangyichao-mango) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java:288]:
```suggestion
                logger.error("exception", e);
                logger.error("delay task execution failure, the task will be executed directly." +
```
Hi,
It will be better to combine these error log to one error log.

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/zk/ZKMasterClient.java [hunk 4]
```diff
@@ -16,6 +16,9 @@
  */
 package org.apache.dolphinscheduler.server.zk;
 
+import java.util.Date;
```
Comment (yangyichao-mango) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/zk/ZKMasterClient.java:19]: Looks like these files not need to be changed, you should revert these changes.
The same with `TaskResponseService.java`, `TaskResponseEvent.java`, `TaskExecuteResponseCommand.java`, `BaseDataSource.java `, `common.properties`.

File: sql/dolphinscheduler-postgre.sql [hunk 5]
```diff
@@ -566,8 +566,10 @@ CREATE TABLE t_ds_task_instance (
   retry_interval int DEFAULT NULL ,
   max_retry_times int DEFAULT NULL ,
   task_instance_priority int DEFAULT NULL ,
-   worker_group varchar(64),
+  worker_group varchar(64),
```
Comment (yangyichao-mango) [inline sql/dolphinscheduler-postgre.sql:569]: You can add the sql updater in this directory.
image

File: dolphinscheduler-remote/src/main/java/org/apache/dolphinscheduler/remote/codec/NettyDecoder.java (hunk 1 above)
Comment (gabrywu) [reply dolphinscheduler-remote/src/main/java/org/apache/dolphinscheduler/remote/codec/NettyDecoder.java]: Of course, you can add a break at the last case

**Comment by vanilla111** [comment]
If a task can be retried after failure and has been delayed enough time, it will no longer enter the state of delayed execution. ------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;&quot;bao&amp;nbsp;liang&quot;<notifications@github.com&gt;
发送时间:&nbsp;2020年8月14日(星期五) 上午10:11
收件人:&nbsp;&quot;apache/incubator-dolphinscheduler&quot;<incubator-dolphinscheduler@noreply.github.com&gt;;
抄送:&nbsp;&quot;vanilla111&quot;<1115690319@qq.com&gt;;&quot;State change&quot;<state_change@noreply.github.com&gt;;
主题:&nbsp;Re: [apache/incubator-dolphinscheduler] [Feature][server,api,common,dao,remote] delay execution of tasks and improve some designs (#3427)

File: dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/entity/TaskInstance.java [hunk 6]
```diff
@@ -298,6 +305,14 @@ public void setState(ExecutionStatus state) {
         this.state = state;
     }
 
+    public Date getFirstSubmitTime() {
+        return firstSubmitTime;
```
[the remark this answers is not in the document]
Comment (vanilla111) [reply dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/entity/TaskInstance.java:309]: If a task can be retried after failure and has been delayed enough time, it will no longer enter the state of delayed execution. This attribute is used to achieve this feature.

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java [hunk 7]
```diff
@@ -256,4 +275,59 @@ private void downloadResource(String execLocalPath,
...
+    /**
+     * delay execution if needed
+     */
+    private void delayExecutionIfNeeded() {
+        long remainTime = DateUtils.getRemainTime(taskExecutionContext.getFirstSubmitTime(),
+                taskExecutionContext.getDelayTime() * 60L);
+        logger.info("delay execution time: {} s", remainTime < 0 ? 0 : remainTime);
+        if (remainTime > 0) {
+            try {
+                Thread.sleep(remainTime * Constants.SLEEP_TIME_MILLIS);
```
Comment (yangyichao-mango) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java:288]: Hi,
After this PR is merged, we can further talk about the optimize of the delayed task execution. The current implementation method is `Thread.sleep()`。

I personally understand that many scenarios that need to use delay should be tens of minutes or several hours. In this scenario, task threads will be occupied all the time and will not be released, which will consume a lot of resources. After this PR is merged, we can talk about the optimization. For example, we can put the delayed tasks into a deferred execution pool and have a separate thread to scan whether they can be executed, and take out the tasks to be executed.

---------------

等这个pr合了之后，我们可以再讨论下把延迟的任务执行优化下，现在的实现方式是`Thread.sleep()`。

我个人理解，很多需要用到延迟的场景应该都是几十分钟或者几个小时的延迟，这种场景下任务线程会一直被占用，得不到释放，会对资源消耗很大。之后可以讨论下是不是可以把延迟执行的任务放在一个延迟执行池里，单独有一个线程去扫描是否可以执行，将需要执行的任务拿出来进行执行。

If you have any suggestions or questions, welcome to put forward~

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/entity/TaskExecutionContext.java [hunk 8]
```diff
@@ -42,6 +43,12 @@
      */
     private String taskName;
 
+    /**
+     *  task first submit time
+     */
+    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss",timezone="GMT+8")
```
Comment (yangyichao-mango) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/entity/TaskExecutionContext.java:49]:
```suggestion
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss", timezone="GMT+8")
```

Comment (yangyichao-mango) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/entity/TaskExecutionContext.java:49]:
```suggestion
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss", timezone="GMT+8")
```

File: dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThreadTest.java [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+        TaskExecuteThread taskExecuteThread = new TaskExecuteThread(taskExecutionContext, taskCallbackService, taskLogger);
+        taskExecuteThread.run();
+
+        Assert.assertEquals(ExecutionStatus.SUCCESS, taskExecutionContext.getCurrentExecutionStatus());
+    }
+
+    @After
+    public void after() throws Exception {
+        //
+    }
```
Comment (yangyichao-mango) [inline dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThreadTest.java:141]: Delete unused function, or you can check whether there is the unreleased resources(eg. Unreleased thread poll) in tests that can be released in this fucntion.

----------
删除没有用到的function，或者可以检查下之前的测试用例中是否有没有释放的资源（比如没有关闭的线程池），可以在这个方法中释放资源。

**Comment by yangyichao-mango** [comment]
Hi~,
After this pr is merged, we also need to talk about the update program about delay execution when user need to update DS from lower version to new version.

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java (hunk 7 above)
Comment (vanilla111) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java:288]: This is indeed a good suggestion, and I have a preliminary idea for its realization. I will start to implement it immediately after I complete another feature.

File: dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThreadTest.java (hunk 9 above)
Comment (vanilla111) [reply dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThreadTest.java:141]: OK, I got it.

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/entity/TaskExecutionContext.java (hunk 8 above)
Comment (vanilla111) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/entity/TaskExecutionContext.java:49]: I will modify it now.
